Honor 'IGNORE_FILES' in StaticGenerator._copy_paths (#1692).

zackw opened the pull request:
generators.StaticGenerator._copy_paths just passes the setting
down; the bulk of the change is in utils.copy, since we want to
have IGNORE_FILES apply at every level of a recursively copied
directory (e.g. the theme static directory).

Since I was overhauling utils.copy anyway, I changed it to use
os.walk, which should be more efficient.

Other users of utils.copy should be unaffected.

avaris commented on pelican/utils.py:
```diff
@@ -273,51 +273,73 @@ def slugify(value, substitutions=()):
+                                    os.path.relpath(src_dir, source_))
+
+            subdirs[:] = (s for s in subdirs if not any(fnmatch.fnmatch(s, i)
+                                                        for i in ignores))
```
Ignoring folders with `IGNORE_FILES` isn't documented and it doesn't match the use of `IGNORE_FILES` in other places. Similar issue with the early `return` on line 301.

zackw commented:
@avaris As there is no `IGNORE_DIRECTORIES`, I am inclined to say that `IGNORE_FILES` should work for both, and if it doesn't include directories elsewhere, *that* should be changed.

avaris commented:
@zackw sure, that's reasonable. If you include that in your PR and update documentation, that would be great. But, ignoring directories in one place but not in other places isn't consistent.

zackw commented:
@avaris Sorry for the long delay getting back to you.  As you suggested, I have made `IGNORE_FILES` apply to directories in the one other place that it's used, and updated the documentation.  I have also split the branch into three commits to clarify what piece does what.

avaris commented:
@zackw looks good, :+1:. Thank you for the work.

justinmayer commented:
Thanks, @zackw!
